File: plugins/collector/event/apps/container/containerd.go
package container

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
)

type Containerd struct {
	version string
	reg     *regexp.Regexp
}

func (Containerd) Name() string { return "containerd" }

func (Containerd) Type() string { return "container" }

func (d Containerd) Version() string { return d.version }

func (Containerd) Match(p *process.Process) bool { return p.Name == "containerd" }

func (n *Containerd) Run(p *process.Process) (m map[string]string, err error) {
	if n.reg == nil {
		n.reg = regexp.MustCompile(`(\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "--version")
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	n.version = str
	return
}

func init() { apps.Regist(&Containerd{}) }

File: plugins/collector/src/event/apps/container/containerd.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Containerd {
    version: String,
}

impl Containerd {
    pub fn new() -> Self { Containerd { version: String::new() } }
}

impl IApp for Containerd {
    fn name(&self)     -> &'static str { "containerd" }
    fn app_type(&self) -> &'static str { "container" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "containerd" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "containerd github.com/containerd/containerd v1.6.8 …"
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
